pandas/core/frame.py

```diff
@@ -716,24 +716,24 @@ def _repr_html_(self):
     def to_string(
         self,
         buf=None,
-        columns=None,
-        col_space=None,
-        header=True,
-        index=True,
-        na_rep="NaN",
-        formatters=None,
-        float_format=None,
-        sparsify=None,
-        index_names=True,
-        justify=None,
-        max_rows=None,
-        min_rows=None,
-        max_cols=None,
-        show_dimensions=False,
-        decimal=".",
-        line_width=None,
-        max_colwidth=None,
-    ):
+        columns: Optional[List[str]] = None,
+        col_space: Optional[Union[str, int]] = None,
+        header: Union[bool, List[str]] = True,
```

Comment:
for `to_string` only bool is permitted?

Reply:
although maybe the docs are ambiguous here..

```
header : bool, optional
   Write out the column names. If a list of strings is given, it is assumed to be aliases for the column names.
```

Reply:
List of `str` for header works as expected. 
> header : bool, optional

Maybe this part the docs needs to be updated.

Reply:
The list must be exactly the same size as number of columns in df.

Reply:
> Maybe this part the docs needs to be updated.

yes. we should try to keep the type annotations and api docs in sync.

> The list must be exactly the same size as number of columns in df.

not sure if there is a way to represent this with typing atm. this constraint should probably be mentioned in the docs. also, I notice a sequence is accepted here, not just a list.

although we should probably not change that as i think Series could be considered a sequence, but..

```
>>> import pandas as pd
>>> pd.__version__
'0.26.0.dev0+350.gad9fe5d7a'
>>> ser = pd.Series(list("cd"))
>>> hasattr(ser, "__iter__")
True
>>>
>>> [x for x in ser]
['c', 'd']
>>>
>>> hasattr(ser, "__len__")
True
>>>
>>> len(ser)
2
>>>
>>> df = pd.DataFrame({'A': [1,2,3], 'B': [4,5,6]})
>>> df
   A  B
0  1  4
1  2  5
2  3  6
>>> print(df.to_string(header=ser))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\frame.py", line 774, in to_string
    return formatter.to_string(buf=buf)
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\io\formats\format.py", line 879, in to_string
    return self.get_result(buf=buf)
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\io\formats\format.py", line 511, in get_result
    self.write_result(buf=f)
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\io\formats\format.py", line 794, in write_result
    strcols = self._to_str_columns()
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\io\formats\format.py", line 690, in _to_str_columns
    str_index = self._get_formatted_index(frame)
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\io\formats\format.py", line 1040, in _get_formatted_index
    if self.header:
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\generic.py", line 1540, in __nonzero__
    self.__class__.__name__
ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().
>>>
```

so should probably leave it as list, and give it some more thought.

Reply:
Any iterable with `__len__` (tuple, set etc.) is accepted. 

    >>> df = pd.DataFrame({'A': [1,2,3], 'B': [4,5,6]})
    >>> df
       A  B
    0  1  4
    1  2  5
    2  3  6

    >>> print(df.to_string(header=('X', 'Y')))  # tuple
       X  Y
    0  1  4
    1  2  5
    2  3  6

    >>> print(df.to_string(header={'X', 'Y'}))  # set
       X  Y
    0  1  4
    1  2  5
    2  3  6

    >>> def h():
    ...     yield 'X'
    ...     yield 'Y'
    ... 

    >>> print(df.to_string(header=h()))  # generator
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
      File "/home/vaibhav/.local/lib/python3.6/site-packages/pandas/core/frame.py", line 761, in to_string
        formatter.to_string()
      File "/home/vaibhav/.local/lib/python3.6/site-packages/pandas/io/formats/format.py", line 688, in to_string
        strcols = self._to_str_columns()
      File "/home/vaibhav/.local/lib/python3.6/site-packages/pandas/io/formats/format.py", line 603, in _to_str_columns
        if len(self.header) != len(self.columns):
    TypeError: object of type 'generator' has no len()

    >>> print(df.to_string(header=tuple(h())))  # works now, as expected.
       X  Y
    0  1  4
    1  2  5
    2  3  6

I think the proper type for `Iterables` with len is `Collection`, but it was added in python 3.6

Reply:
although, the types for the public api should probably be as permissible as possible, i think we should just leave the type for header as list for now to avoid scope creep.

Reply:
lets see what @WillAyd thinks about this.

Reply:
How about `Sequence`? I think `List` is a little too restrictive

Reply:
Would be nice to update doc as part of this as well

Reply:
If this supports generators then maybe `Iterable` is better than `Sequence`. If not stick to `Sequence`

Reply:
generated needs to evaluated into something like list, tuple https://github.com/pandas-dev/pandas/pull/28453#discussion_r324647644. Just a generator instance will not work, as it has no `__len__`

Reply:
Changed it to `Sequence`.